[android] Reduce size of apk

Jean-BaptisteC opened the pull request:
This PR add resConfig in build.gradle to remove unused strings in librairies.
https://github.com/organicmaps/organicmaps/pull/3584#issuecomment-1276473507

APK have lost 600Ko ~

This PR works with per-app language.

rtsisyk commented on android/build.gradle:
```diff
@@ -161,6 +161,7 @@ android {
     buildConfigField 'String', 'SUPPORT_MAIL', '"android@organicmaps.app"'
     buildConfigField 'String', 'REVIEW_URL', '"https://organicmaps.app/donate/"'
     buildConfigField 'int', 'RATING_THRESHOLD', '5'
+    resConfigs "en", "ar", "be", "bg", "ca", "cs", "da", "de", "el", "en_GB", "es", "es_MX", "eu", "fa", "fi", "fr", "fr_CA", "hu", "in", "it", "iw", "ja", "ko", "mr", "nb", "nl", "pl", "pt", "pt_BR", "ro", "ru", "sk", "sv", "sw", "th", "tr", "uk", "vi", "zh", "zh_HK", "zh_MO", "zh_TW"
```
zh_TW or zh-rTW?

Jean-BaptisteC replied:
zh_TW take in zh-rTW during compilation.
I have use per-app language and strings in values_fr-rCA.xml are available in app.

rtsisyk replied:
OK, I tested that `en-rGB` works

biodranik replied:
1. Please use single quotes for constants without vars inside them.
2. Is it possible to load them from a config file?
3. Is there a way to automatically calculate and set these languages based on our real translations?
4. Please update the documentation so that when a new language is added, this line should be updated too.

Jean-BaptisteC replied:
> 1. Please use single quotes for constants without vars inside them.

Yes :)

>  2. Is it possible to load them from a config file?

Yes, we can add resConfig in [gradle.properties](https://github.com/organicmaps/organicmaps/blob/master/android/gradle.properties) Check [this](https://medium.com/livefront/improve-android-build-times-by-using-resconfigs-99177a3c6d90)

> 3. Is there a way to automatically calculate and set these languages based on our real translations?

Not, you need to specify which language do you want integrate in Android build

> 4. Please update the documentation so that when a new language is added, this line should be updated too.

Do you want i add new section in [TRANSLATIONS.md](https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md) ? Or just specify a new line in Translation files section with link to build.gradle with description `List of languages integrate in Android build`

rtsisyk replied:
> > 1. Please use single quotes for constants without vars inside them.
> 
> Yes :)

Thanks! This is not documented, but here is a strong belief that single quotes are easier to read than double quota, which is a case in Bash. I filed a follow-up ticket to research https://github.com/organicmaps/organicmaps/issues/3684.

> 
> > 2. Is it possible to load them from a config file?
> 
> Yes, we can add resConfig in [gradle.properties](https://github.com/organicmaps/organicmaps/blob/master/android/gradle.properties) Check [this](https://medium.com/livefront/improve-android-build-times-by-using-resconfigs-99177a3c6d90)

I don't see any difference between having this list in `gradle.properties` vs `build.gradle`. It doesn't change things significantly. Both options look like the same for me. 

@biodranik do you have any better ideas? 

> 
> > 3. Is there a way to automatically calculate and set these languages based on our real translations?
> 
> Not, you need to specify which language do you want integrate in Android build
> 

Generating this list dynamically based on `values-*` directory sounds complicated to me. It is just not worth it.

> > 4. Please update the documentation so that when a new language is added, this line should be updated too.
> 
> Do you want i add new section in [TRANSLATIONS.md](https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md) ? Or just specify a new line in Translation files section with link to build.gradle with description `List of languages integrate in Android build`

I don't see a good place where you can add a comment in translation files (string.txt). TRANSLATIONS.md should be fine.

biodranik replied:
I have a better idea. If languages can be loaded from file, let's update tools/unix/generate_localizations.sh to autogenerate this file with the list of all supported languages, for example, in strings.txt.

rtsisyk replied:
We don't add new languages every day. Can we keep it simple?

biodranik replied:
It is very easy to forget it. With the help of the community, I expect many languages to be added. Let's do it properly from the start. Once and for all.

biodranik replied:
This one-liner can be inserted into generate_localizations.sh to autogenerate the include config for android:
```bash
cat data/strings/strings.txt| sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/'\1',/p" | sort -u | tr '\n' ' ' | sed -E 's/, $//'
'ar', 'be', 'bg', 'ca', 'cs', 'da', 'de', 'el', 'en', 'en-GB', 'es', 'es-MX', 'et', 'eu', 'fa', 'fi', 'fr', 'fr-CA', 'he', 'hi', 'hu', 'id', 'in', 'it', 'ja', 'ko', 'mr', 'nb', 'ne', 'nl', 'pl', 'pt', 'pt-BR', 'ro', 'ru', 'sk', 'sv', 'sw', 'th', 'tr', 'uk', 'vi', 'zh-Hans', 'zh-Hant'
```

Jean-BaptisteC replied:
Can you give more information about what doing these commands, to better understand what i need to add in build.gradle.
Is it ok, if i add your commands just in end file before rm in .generate_localizations.sh ?

biodranik replied:
The first line is the command.
The second line is its output.

You need to include a file with these langs into build.gradle instead of hard-coding languages there, and generate this file automatically from the mentioned script.

Jean-BaptisteC replied:
- List of languages is include in gradle.properties
- resConfig call gradle.properties (It's more simple solution, i have try with generate txt file, but i didn't have find how import content of file and use in resConfig)
List of languages is more longer and when i try to build, compilation fail, can you check ?
More logs here :https://github.com/Jean-BaptisteC/organicmaps/actions/runs/3450378278/jobs/5758934308

biodranik commented on android/build.gradle:
```diff
@@ -161,6 +161,7 @@ android {
     buildConfigField 'String', 'SUPPORT_MAIL', '"android@organicmaps.app"'
     buildConfigField 'String', 'REVIEW_URL', '"https://organicmaps.app/donate/"'
     buildConfigField 'int', 'RATING_THRESHOLD', '5'
+    resConfigs project.ext.listLanguages
```
I would name it supportedLocalizations

biodranik commented on tools/unix/generate_localizations.sh:
```diff
@@ -66,4 +66,8 @@ cat "$STRINGS_PATH"/{strings,types_strings}.txt> "$MERGED_FILE"
 
+# Generate list of languages and add list in gradle.properties to be used in build.gradle in resConfig
+RES_CONFIG=$(sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/'\1',/p" data/strings/strings.txt | sort -u | tr '\n' ' ' | sed -E 's/, $//')
+echo listLanguages="$RES_CONFIG" >> "$OMIM_PATH/android/gradle.properties"
```
It will add the line _each time_ the script is run. The robust implementation should _add or replace existing_ line if it's already there.

biodranik commented on tools/unix/generate_localizations.sh:
```diff
@@ -66,4 +66,8 @@ cat "$STRINGS_PATH"/{strings,types_strings}.txt> "$MERGED_FILE"
 "$TWINE" generate-all-localization-files --format jquery "$OMIM_PATH/data/strings/sound.txt" "$OMIM_PATH/data/sound-strings/"
 
+# Generate list of languages and add list in gradle.properties to be used in build.gradle in resConfig
+RES_CONFIG=$(sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/'\1',/p" data/strings/strings.txt | sort -u | tr '\n' ' ' | sed -E 's/, $//')
```
```suggestion
SUPPORTED_LOCALIZATIONS=$(sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/'\1',/p" "${OMIM_PATH}/data/strings/strings.txt" | sort -u | tr '\n' ' ' | sed -E 's/, $//')
```

biodranik commented:
Looks like gradle can't read list of values from gradle.properties file. You may need to change the script to generate string like this: `en,be,ru,uk` and then assign it as an array like this: https://stackoverflow.com/a/49614449

Jean-BaptisteC commented:
I have fixed problems during compilation (space is not allow by gradle, and we need to - replace by _) and i have already add list in gradle.properties.
I have try to fixed with add new commands in pipeline but linux say too many arguments.
I have try with tr  in new line (`"$SUPPORTED_LOCALIZATIONS"=$(echo "$SUPPORTED_LOCALIZATIONS" | tr '-' '_' | tr -d " ")`) but i have others errors :(

I missing knowledge about linux scripting, finish PR can be take a lot of time.

biodranik commented on tools/unix/generate_localizations.sh:
```diff
@@ -66,4 +66,12 @@ cat "$STRINGS_PATH"/{strings,types_strings}.txt> "$MERGED_FILE"
 
+# Generate list of languages and add list in gradle.properties to be used in build.gradle in resConfig
+SUPPORTED_LOCALIZATIONS=$(sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/\1,/p" "${OMIM_PATH}/data/strings/strings.txt" | sort -u | tr '\n' ' ' | sed -E 's/, $//')
+SUPPORTED_LOCALIZATIONS=$(echo listLanguages="$SUPPORTED_LOCALIZATIONS" | sed 's/[[:space:]]//g' | tr '-' '_')
```
You definitely can generate the necessary list format in the above command, instead of fixing it here. Or no?

Jean-BaptisteC replied:
Unix tell me too many arguments

Jean-BaptisteC replied:
Fixed :)

Jean-BaptisteC commented:
I have finish script bash:

- I have replace `-` by `_`
- I have remove space between languages and remove quotes
- I have update list of languages in gradle.properties with latest languages from master
- I have add condition to update gradle.properties only if languages have changed (To compare, i have add new variable with TEST like name, i don't have find good name for this)

biodranik commented on tools/unix/generate_localizations.sh:
```diff
@@ -66,4 +66,11 @@ cat "$STRINGS_PATH"/{strings,types_strings}.txt> "$MERGED_FILE"
 "$TWINE" generate-all-localization-files --format jquery "$OMIM_PATH/data/strings/sound.txt" "$OMIM_PATH/data/sound-strings/"
 
+# Generate list of languages and add list in gradle.properties to be used in build.gradle in resConfig
+SUPPORTED_LOCALIZATIONS="supportedLocalizations="$(sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/\1,/p" "${OMIM_PATH}/data/strings/strings.txt" | sort -u | tr '\n' ' ' | sed -E 's/, $//' | sed 's/[[:space:]]//g' | tr '-' '_')
```
The shorter version is `sed -nEe "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/\1/p" "data/strings/strings.txt" | sort -u | tr '\n' ',' | sed -e 's/-/_/g'`

Jean-BaptisteC replied:
This version add comma after last element :(

biodranik replied:
Just add `-e 's/,$//'` to the last sed invocation.

biodranik commented on tools/unix/generate_localizations.sh:
```diff
@@ -66,4 +66,11 @@ cat "$STRINGS_PATH"/{strings,types_strings}.txt> "$MERGED_FILE"
+SUPPORTED_LOCALIZATIONS="supportedLocalizations="$(sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/\1,/p" "${OMIM_PATH}/data/strings/strings.txt" | sort -u | tr '\n' ' ' | sed -E 's/, $//' | sed 's/[[:space:]]//g' | tr '-' '_')
+TEST=$(grep supportedLocalizations "$OMIM_PATH/android/gradle.properties")
+if [ "$SUPPORTED_LOCALIZATIONS" != "$TEST" ] ; then
+sed -i "s/$TEST/$SUPPORTED_LOCALIZATIONS/" "$OMIM_PATH/android/gradle.properties"
```
No indent.

biodranik replied:
sed -i won't work on Mac (it requires `-i ''` syntax to avoid creating a backup file).

Does it make sense to load gradle.properties content into a variable, and do all replacements there, if necessary, and then overwrite the original file?

biodranik commented on tools/unix/generate_localizations.sh:
```diff
@@ -66,4 +66,11 @@ cat "$STRINGS_PATH"/{strings,types_strings}.txt> "$MERGED_FILE"
+# Generate list of languages and add list in gradle.properties to be used in build.gradle in resConfig
+SUPPORTED_LOCALIZATIONS="supportedLocalizations="$(sed -nE "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/\1,/p" "${OMIM_PATH}/data/strings/strings.txt" | sort -u | tr '\n' ' ' | sed -E 's/, $//' | sed 's/[[:space:]]//g' | tr '-' '_')
+TEST=$(grep supportedLocalizations "$OMIM_PATH/android/gradle.properties")
+if [ "$SUPPORTED_LOCALIZATIONS" != "$TEST" ] ; then
```
```suggestion
if [ "$SUPPORTED_LOCALIZATIONS" != "$TEST" ]; then
```

biodranik reviewed (changes requested):
Will it work if gradle.properties is empty? Will sed in L73 properly add it at the end?

Jean-BaptisteC commented:
> Will it work if gradle.properties is empty? Will sed in L73 properly add it at the end?

Do you want i add condition with grep to check if line existed in gradle.properties ?

biodranik commented on android/gradle.properties:
```diff
@@ -8,3 +8,4 @@ org.gradle.jvmargs=-Xmx1024m -Xms256m
 android.useAndroidX=true
 android.enableJetifier=true
 android.native.buildOutput=verbose
+supportedLocalizations=ar,be,bg,ca,cs,da,de,el,en,en_GB,es,es_MX,et,eu,fa,fi,fr,fr_CA,he,hi,hu,id,in,it,ja,ko,mr,nb,ne,nl,pl,pt,pt_BR,ro,ru,sk,sv,sw,th,tr,uk,vi,zh_Hans,zh_Hant
```
Can we add a comment above? Like
`# Autogenerated by tools/unix/generate_localizations.sh`
?

Jean-BaptisteC commented:
I have fixed all things on my computer, but i can't rebase my PR because Github Desktop have generated merge branch commit can you help me ?
I have added commentary on gradle.properties and fixed script

```
# Generate list of languages and add list in gradle.properties to be used in build.gradle in resConfig
SUPPORTED_LOCALIZATIONS="supportedLocalizations="$(sed -nEe "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/\1/p" "data/strings/strings.txt" | sort -u | tr '\n' ',' | sed -e 's/-/_/g' -e 's/,$//')
if [ "$SUPPORTED_LOCALIZATIONS" != "$(grep listLanguages "$OMIM_PATH/android/gradle.properties")" ]; then
	sed -i 's/supportedLocalizations.*/'"${SUPPORTED_LOCALIZATIONS}"'/' "$OMIM_PATH/android/gradle.properties"
fi
```

biodranik commented:
You should learn two important things we actively use here in OM development:
1. Rebase
2. Resolving merge conflicts (they can be easily fixed in any text editor).

I've fixed that conflict for you using Github web interface.

Jean-BaptisteC commented:
I have planned to rebase branch but i have need to tests on other branch to be sure i don't break config branch, because i always use Github Desktop, i have never use git CLI.

Thanks to have resolve conflicts :)
